reps/2023-02-01-wasm-on-ray.md

```diff
@@ -0,0 +1,253 @@
+
+### Other Languages Support
+
+#### Rust
+It is also possible to support other languages that can be compiled to WASM. For example in Rust, you can declare the Ray APIs you want to import in the following way.
+
+```rust
+#[link(wasm_import_module = "ray")] 
+extern "C" {
+  #[link_name="call"]
+  fn rcall(ptr: *const u8, ...) -> u32;
+}
+
+```
+
+Then you can use the `rcall` function to call the remote function. The `ptr` is the function pointer. The following arguments are the arguments of the remote function.
+
+
+#### Other Languages
+
+To use Ray APIs in JavaScript, the Ray APIs need to be exported to the Javascript runtime environment. For example, if the user is using SpiderMonkey, the Ray APIs can be exported to SpiderMonkey and then used in JavaScript.
+
+Futhermore, the new Component Model standard is supposed to be ready very soon. It will provide a standard way to import and export components. We can use the new Component Model standard to declare Ray APIs in WASM so that it will be easier to use Ray APIs in different languages.
+
+
+## Pros and Cons
+Compared with directly supporting other languages, WASM support in Ray has the following pros and cons.
+
+### Pros:
+1. It is easy to support new languages. For example, if we want to support a new language, we only need to compile the language to WASM and then use the WASM support in Ray to run the WASM binary.
+2. Offer the ability to run Ray-As-A-Service. With WASM enabled in Ray, it is possible for users to use untrusted third-party WASM binaries in Ray. As WASM supports a sandboxed environment, the activities of the untrusted third-party WASM binaries can do is limited.
+3. Lightweight Workers. WASM is only 10% the size of Python for comparable tasks. This means that the WASM worker is much smaller than the Python worker. WASM is lightweight and fast to load. It is possible to load the WASM binary in a few milliseconds. Compared to Python, it is much faster to load the WASM binary.
+4. Architecture Agnostic. WASM provides an abstract layer on top of the hardware. It is possible to support different hardware architectures in the same WASM binary.
+5. Easy to support different versions of the same language. For example, if we want to support different versions of JavaScript, we only need to compile different versions of JavaScript to WASM and then use the WASM support in Ray to run the WASM binary.
+
+### Cons:
+1. It is not easy to support all Ray APIs in WASM. For example, there is a concept of `Actor` in Ray. `Actor` is a special type of task that can maintain state. However, WASM is a collection of functions, a stack, and a heap. It is not a simple task to map the concept of `Actor` to WASM. Therefore, we will not support `Actor` in WASM in the first version of WASM support in Ray.
```

Comment:
It's a significant con.

Reply:
I think this part can help us review some of the Ray core APIs so that we can improve the Ray core interface. For the MVP, I think we cannot yet support Actor. With more brainstorming and design sessions, I think there will be a solution. I didn't see any technical challenges so far.

Reply:
@jjyao there are several things here:

- Build workers
    - support functions (first step)
    - support actors (third step)
- Define middle layer apis (second step)

The difficulty comes from ray side. To be more specific, ray doesn't have good developer api to add new languages. And this is one of the goal of this project, to figure out what might be good api we should define.

Reply:
Figuring out a good middle layer for multi-language support is indeed very important (IMO, more important than supporting WASM itself), are we going to have a design for that?

Reply:
@jjyao I know what you mean. The fact here is that WASM worker support is critical to @wilsonwang371 and middle layer support is important for Ray. Given that we don't know what's needed, it's a good chance we'll learn it from this work by doing first step (support functions).

TBH, depend on the resources we have it might or might not happen. But at lease we'll understand the requirement better after finishing step 1.

Reply:
totally agreed. Actually, we are also willing to contribute to the middle layer API.